### PR TITLE
ENH: add dtype option to numpy.lib.function_base.cov and corrcoef

### LINKED ISSUE #17395: dtype control for corrcoef
## Feature

I am computing a cross-correlation matrix of 150K * 150K matrix of float32.

Numpy insists on making the results dtype be float64. This causes the amount of memory to be "too much".

I would debate that default, but changing it would probably be too disruptive (not backward compatible).

However, how about a dtype parameter for corrcoef?

In general I would have liked to see a dtype parameter "everywhere" to override the float64 default...

### PR DESCRIPTION
**Pull request by lschwetlick**
I tried to implement the changes suggested in [Issue #17395](https://github.com/numpy/numpy/issues/17395#issue-711108059), by adding a dtype keyword to the cov and corrcoef functions in numpy/lib/function_base.py. I also wrote tests for the changes.

Closes #17395

### CODE REVIEW

File: numpy/lib/function_base.py
Comment (eric-wieser) [inline numpy/lib/function_base.py:2411]: You need to leave the `ValueError` outside the `dtype` if
```suggestion
    if y is not None:
        y = np.asarray(y)
        if y.ndim > 2:
            raise ValueError("y has more than 2 dimensions")

    if dtype is None:
        if y is None:
            dtype = np.result_type(m, np.float64)
        else:
            dtype = np.result_type(m, y, np.float64)
```

File: numpy/lib/tests/test_function_base.py
Comment (eric-wieser) [inline numpy/lib/tests/test_function_base.py:2030]: Rather than skipping, just use the names which always exist:
```suggestion
    @pytest.mark.parametrize("test_type", [np.half, np.single, np.double, np.longdouble])
    def test_cov_dtype(self, test_type):
```

File: numpy/lib/function_base.py
Comment (eric-wieser) [inline numpy/lib/function_base.py:2277]: Let's make this keyword-only. (other arguments aren't done that way because the syntax didn't exist when this function was written)
```suggestion
        aweights=None, *, dtype=None):
```

File: numpy/lib/tests/test_function_base.py
Comment (lschwetlick) [reply numpy/lib/tests/test_function_base.py:2030]: Nice, I wasnt aware of those. Thats much better, thanks!

Comment (eric-wieser) [reply numpy/lib/tests/test_function_base.py]: The docs aren't clear about that yet, so it's understandable you missed them - I'm updating them in #17331. The link at the top of that PR will take you to the new version.

**Review by eric-wieser — COMMENTED** [review]
This generally looks good, thanks for the contribution. Two more things:

* The `dtype` argument needs documenting, and should include `.. versionadded:: 1.20.0`
* We should maybe add a release note in the `upcoming_changes` folder.

**Comment by lschwetlick** [comment]
Sure, I can certainly do that. I left it out initially to see whether there was interest in actually merging the PR. Thank you for reviewing so quickly @eric-wieser !

File: numpy/lib/function_base.py
Comment (eric-wieser) [inline numpy/lib/function_base.py:2329]:
```suggestion
        Data-type of the result. By default, the return data-type is `numpy.float64`.
```
which will produce a link

**Review by eric-wieser — APPROVED** [review]
This probably doesn't need to hit the mailing list, but I'll leave it for another maintainer to give a second opinion on. Thanks for the rapid iteration.

**Comment by lschwetlick** [comment]
Happy to help! 

Its a bit silly maybe but if its no trouble, could you add the label "hacktoberfest-accepted" to the PR? [They changed the rules to prevent spam](https://hacktoberfest.digitalocean.com/hacktoberfest-update).

**Comment by eric-wieser** [comment]
@lschwetlick, there's no need for me to do that, apparently, my approval above is enough.

What _does_ matter is if the repo has a `hacktoberfest` topic, which it currently does not. I'd be happy to add it to make this PR count, but I worry that we might draw span attention as well.

**Review by rossbar — APPROVED** [review]
LGTM as well, seems like a nice improvement. Thanks @lschwetlick !

File: numpy/lib/tests/test_function_base.py
Comment (eric-wieser) [inline numpy/lib/tests/test_function_base.py:2030]: @rossbar: I'd probably be tempted to have left this as `assert test_type == res_dtype` - now that we have pytest, there's little reason to use our test functions for anything other than arrays.

That's not to say it's worth changing it back either, just something to consider for next time :)

Comment (rossbar) [reply numpy/lib/tests/test_function_base.py:2030]: Thanks for the heads up!

File: numpy/lib/function_base.py
Comment (seberg) [inline numpy/lib/function_base.py:2330]: I am happy if we put this in. Technically, the default has at least `numpy.float64` precision (mostly, it could be `object` or `longdouble`).  I am a bit unsure if explaining that makes it just harder to understand though.

**Comment by eric-wieser** [comment]
Added the hacktoberfest-accepted label, since I think @mattip said that was preferable to the label. Apologies if it's too late to make a difference!

**Comment by lschwetlick** [comment]
Nice, thank you @eric-wieser !

**Pull request merged**
